### PR TITLE
change shiftdata so 1 and 2 points work, add tests.. (#213)

### PR DESCRIPTION
**Pull request by guziy**
A way to make shiftdata work with one or 2 points, this is probably addressing all the issues of scatter and plot with less than 3 points...

### CODE REVIEW

**Comment by WeatherGod** [comment]
Actually, this would work. Could you do the same for the 1D case below as well?

Also, this branch has conflicts, for some reason...

**Comment by WeatherGod** [comment]
sorry, I mean the 2D case above

**Comment by guziy** [comment]
I've added a test and the change for the 2d case, but am not sure what's causing conflicts...

Sorry about the conflicts, I've started off an older master.. Now I merged it and reran the tests.

File: lib/mpl_toolkits/basemap/__init__.py [hunk 1]
```diff
@@ -4777,10 +4777,15 @@ def shiftdata(self,lonsin,datain=None,lon_0=None):
...
-            londiff_sort = np.sort(londiff)
-            thresh = 360.-londiff_sort[-2]
-            itemindex = nlons-np.where(londiff>=thresh)[0]
+            if nlons > 1:
+                londiff = np.abs(lonsin1[0:-1]-lonsin1[1:])
+                londiff_sort = np.sort(londiff)
+                thresh = 360.-londiff_sort[-2] if nlons > 2 else 360.-londiff_sort[-1]
+                itemindex = nlons-np.where(londiff>=thresh)[0]
+            else:
+                lonsin[0, :] = lonsin1
```
Comment (WeatherGod) [inline lib/mpl_toolkits/basemap/__init__.py]: What's the point of this line?

Comment (guziy) [inline lib/mpl_toolkits/basemap/__init__.py]: It shifts one point to the projection region.

In 2d case lonsin is copied to lonsin1 and then lonsin1 is shifted.. Then
lonsin is also shifted below if itemindex is not 0. But when there is only
one point itemindex is zero, so I had to copy back the shifted lonsin1...

2015-10-26 15:51 GMT-04:00 Benjamin Root <notifications@github.com>:

> In lib/mpl_toolkits/basemap/__init__.py
> <https://github.com/matplotlib/basemap/pull/215#discussion_r43043356>:
>
> > @@ -4777,10 +4777,15 @@ def shiftdata(self,lonsin,datain=None,lon_0=None):
> >              lonsin1 = lonsin[0,:]
> >              lonsin1 = np.where(lonsin1 > lon_0+180, lonsin1-360 ,lonsin1)
> >              lonsin1 = np.where(lonsin1 < lon_0-180, lonsin1+360 ,lonsin1)
> > -            londiff = np.abs(lonsin1[0:-1]-lonsin1[1:])
> > -            londiff_sort = np.sort(londiff)
> > -            thresh = 360.-londiff_sort[-2]
> > -            itemindex = nlons-np.where(londiff>=thresh)[0]
> > +            if nlons > 1:
> > +                londiff = np.abs(lonsin1[0:-1]-lonsin1[1:])
> > +                londiff_sort = np.sort(londiff)
> > +                thresh = 360.-londiff_sort[-2] if nlons > 2 else 360.-londiff_sort[-1]
> > +                itemindex = nlons-np.where(londiff>=thresh)[0]
> > +            else:
> > +                lonsin[0, :] = lonsin1
>
> What's the point of this line?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/basemap/pull/215/files#r43043356>.
>



-- 
Sasha

**Comment by guziy** [comment]
@WeatherGod Now you got me thinking, the shiftdata modifies the input array of longitudes, which might be a problem if someone like me would not expect it))

File: lib/mpl_toolkits/basemap/test.py [hunk 2]
```diff
@@ -96,6 +96,59 @@ def test_no_cyc2(self):
...
+        return np.meshgrid(lons1d, lats)[0]
+
+    def test_2_points_should_work(self):
+        """
+        Shiftdata should work with 2 points
+        """
+        bm = Basemap(llcrnrlon=0, llcrnrlat=-80, urcrnrlon=360, urcrnrlat=80, projection='mill')
+
+        lons_expected = [10, 15, 20]
+        lonsout = bm.shiftdata(lons_expected.copy())
```
Comment (WeatherGod) [inline lib/mpl_toolkits/basemap/test.py]: lists do not have copy() methods.

By the way, the testing mechanism isn't picking up these tests. I am about to submit a PR that will fix that.

**Comment by guziy** [comment]
In python3 they do)) Sorry I am using python3, do not have python2 to check this... There is a method that picks them all, are you going to register these as well? I am developing and running tests in Pycharm, which allows to just select tests to run)) Sorry about the omissions..

Cheers

File: lib/mpl_toolkits/basemap/test.py (hunk 2 above)
Comment (guziy) [inline lib/mpl_toolkits/basemap/test.py]: In python3 they do)) Sorry I am using python3, do not have python2 to check this... There is a method that picks them all, are you going to register these as well? I am developing and running tests in Pycharm, which allows to just select tests to run)) Sorry about the omissions..

Cheers

Comment (WeatherGod) [inline lib/mpl_toolkits/basemap/test.py]: Don't worry about handling the selection of tests. Just fix this copy() issue, and I'll merge this. Then rebase my PR to get basic unittest stuff working. I am probably gonna have to overhaul the unittesting at some point anyway as the "tests" are all over the place.

**Pull request merged**